cpp/src/interop/arrow_to_cudf.cu

```diff
@@ -0,0 +1,302 @@
+                        ? std::move(child.first)
+                        : std::make_unique<column>(child.second, stream, mr);
+
+  auto num_rows = offsets_column->size() - 1;
+  auto out_col  = make_lists_column(num_rows,
+                                   std::move(offsets_column),
+                                   std::move(child_column),
+                                   UNKNOWN_NULL_COUNT,
+                                   std::move(*get_mask_buffer(array, mr, stream)));
+
+  return std::make_pair<std::unique_ptr<column>, column_view>(
+    std::move(out_col),
+    slice(out_col->view(),
+          static_cast<size_type>(array.offset()),
+          static_cast<size_type>(array.offset() + array.length())));
+}
+
+std::pair<std::unique_ptr<column>, column_view> get_column(arrow::Array const& array,
+                                                           data_type type,
+                                                           bool skip_mask,
+                                                           rmm::mr::device_memory_resource* mr,
+                                                           cudaStream_t stream)
+{
+  return type_dispatcher(type, dispatch_to_cudf_column{}, array, type, skip_mask, mr, stream);
+}
+
+}  // namespace
+
+std::unique_ptr<table> from_arrow(arrow::Table const& input_table,
+                                  rmm::mr::device_memory_resource* mr,
+                                  cudaStream_t stream)
+{
+  std::vector<std::unique_ptr<column>> columns;
+  auto chunked_arrays = input_table.columns();
+  std::transform(chunked_arrays.begin(),
+                 chunked_arrays.end(),
+                 std::back_inserter(columns),
+                 [&mr, &stream](auto const& chunked_array) {
+                   std::vector<std::unique_ptr<column>> concat_columns;
+                   std::vector<column_view> concat_column_views;
+                   auto cudf_type    = arrow_to_cudf_type(chunked_array->type()->id());
+                   auto array_chunks = chunked_array->chunks();
+
+                   transform(
+                     array_chunks.begin(),
+                     array_chunks.end(),
+                     std::back_inserter(concat_columns),
+                     [&cudf_type, &concat_column_views, &mr, &stream](auto const& array_chunk) {
+                       auto col_and_view = get_column(*array_chunk, cudf_type, false, mr, stream);
+                       concat_column_views.emplace_back(col_and_view.second);
+                       return std::move(col_and_view.first);
+                     });
+
+                   if (concat_columns.size() == 1) {
+                     if ((concat_column_views[0].offset() == 0) and
+                         (concat_column_views[0].size() == concat_columns[0]->size())) {
+                       return std::move(concat_columns[0]);
+                     } else {
+                       return std::make_unique<column>(concat_column_views[0]);
```

Comment:
This is making another copy just to account for the `offset`! 

See my above comment to make `get_column` only copy `[offset, size)` from the `arrow::Array`.